src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,688 @@
+									quote_literal_cstr(tenantId))));
+		}
+
+		strcpy_s(attributeToTenant, sizeof(attributeToTenant), tenantId);
+		attributeToTenantStart = clock();
+	}
+	else
+	{
+		strcpy_s(attributeToTenant, sizeof(attributeToTenant), "");
+	}
+
+	/*DetachSegment(); */
+}
+
+
+/*
+ * AnnotateQuery annotates the query with tenant attributes.
+ */
+char *
+AnnotateQuery(char *queryString, char *partitionColumn, int colocationId)
+{
+	if (partitionColumn == NULL)
+	{
+		return queryString;
+	}
+	StringInfo newQuery = makeStringInfo();
+	appendStringInfo(newQuery, ATTRIBUTE_STRING_FORMAT, partitionColumn, colocationId);
+
+	appendStringInfoString(newQuery, queryString);
+
+	return newQuery->data;
+}
+
+
+/*
+ * CitusAttributeToEnd keeps the statistics for the tenant and calls the previously installed end hook
+ * or the standard executor end function.
+ */
+void
+CitusAttributeToEnd(QueryDesc *queryDesc)
+{
+	/*
+	 * At the end of the Executor is the last moment we have to attribute the previous
+	 * attribution to a tenant, if applicable
+	 */
+	AttributeMetricsIfApplicable();
+
+	/* now call in to the previously installed hook, or the standard implementation */
+	if (prev_ExecutorEnd)
+	{
+		prev_ExecutorEnd(queryDesc);
+	}
+	else
+	{
+		standard_ExecutorEnd(queryDesc);
+	}
+}
+
+
+/*
+ * CompareTenantScore is used to sort the tenant statistics by score
+ * in descending order.
+ */
+static int
+CompareTenantScore(const void *leftElement, const void *rightElement)
+{
+	const TenantStats *leftTenant = (const TenantStats *) leftElement;
+	const TenantStats *rightTenant = (const TenantStats *) rightElement;
+
+	if (leftTenant->score > rightTenant->score)
+	{
+		return -1;
+	}
+	else if (leftTenant->score < rightTenant->score)
+	{
+		return 1;
+	}
+	return 0;
+}
+
+
+/*
+ * AttributeMetricsIfApplicable updates the metrics for current tenant's statistics
+ */
+static void
+AttributeMetricsIfApplicable()
+{
+	if (strcmp(attributeToTenant, "") != 0)
+	{
+		clock_t end = { 0 };
+
+		end = clock();
+		time_t queryTime = time(0);
+		double cpu_time_used = ((double) (end - attributeToTenantStart)) / CLOCKS_PER_SEC;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attribute cpu counter (%f) to tenant: %s",
+									cpu_time_used, attributeToTenant)));
+		}
+
+		MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+		LWLockAcquire(&monitor->lock, LW_SHARED);
+
+		int currentTenantIndex = FindTenantStats(monitor);
+
+		if (currentTenantIndex != -1)
+		{
+			TenantStats *tenantStats = &monitor->tenants[currentTenantIndex];
+			LWLockAcquire(&tenantStats->lock, LW_EXCLUSIVE);
+
+			UpdatePeriodsIfNecessary(tenantStats, queryTime);
+			ReduceScoreIfNecessary(tenantStats, queryTime);
+			RecordTenantStats(tenantStats);
+
+			LWLockRelease(&tenantStats->lock);
+		}
+		else
+		{
+			LWLockRelease(&monitor->lock);
+
+			LWLockAcquire(&monitor->lock, LW_EXCLUSIVE);
```

Comment:
we should be careful that this does not happen often for typical CRUD workloads (e.g. distribution column is bigserial), otherwise we get significant contention

Reply:
I'm turning the tracking off by default